npm/nameSpaceController.go

```diff
@@ -0,0 +1,492 @@
+			utilruntime.HandleError(fmt.Errorf("NameSpace '%s' in work queue no longer exists", key))
+			// find the namespace object from a local cache and start cleaning up process (calling cleanUpDeletedPod function)
+			nsKey := util.GetNSNameWithPrefix(key)
+			cachedNs, found := nsc.npMgr.NsMap[nsKey]
+			// if the namespace does not exists, we do not need to clean up process and retry it
+			if !found {
+				return nil
+			}
+
+			// Found the namespace object from NsMap local cache and start cleaning up processes
+			err = nsc.cleanDeletedNamespace(cachedNs.name, cachedNs.LabelsMap)
+			if err != nil {
+				// need to retry this cleaning-up process
+				metrics.SendErrorLogAndMetric(util.NSID, "Error: %v when namespace is not found", err)
+				return fmt.Errorf("Error: %v when namespace is not found\n", err)
+			}
+			return err
+		}
+	}
+
+	if nsObj.DeletionTimestamp != nil || nsObj.DeletionGracePeriodSeconds != nil {
+		return nsc.cleanDeletedNamespace(
+			util.GetNSNameWithPrefix(nsObj.Name),
+			nsObj.Labels,
+		)
+
+	}
+
+	err = nsc.syncUpdateNameSpace(nsObj)
+	// 1. deal with error code and retry this
+	if err != nil {
+		return fmt.Errorf("failed to sync namespace due to  %s", err.Error())
+	}
+
+	return nil
+}
+
+// InitAllNsList syncs all-namespace ipset list.
+func (npMgr *NetworkPolicyManager) InitAllNsList() error {
+	allNs := npMgr.NsMap[util.KubeAllNamespacesFlag]
+	for ns := range npMgr.NsMap {
+		if ns == util.KubeAllNamespacesFlag {
+			continue
+		}
+
+		if err := allNs.IpsMgr.AddToList(util.KubeAllNamespacesFlag, ns); err != nil {
+			metrics.SendErrorLogAndMetric(util.NSID, "[InitAllNsList] Error: failed to add namespace set %s to ipset list %s with err: %v", ns, util.KubeAllNamespacesFlag, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UninitAllNsList cleans all-namespace ipset list.
+func (npMgr *NetworkPolicyManager) UninitAllNsList() error {
+	allNs := npMgr.NsMap[util.KubeAllNamespacesFlag]
+	for ns := range npMgr.NsMap {
+		if ns == util.KubeAllNamespacesFlag {
+			continue
+		}
+
+		if err := allNs.IpsMgr.DeleteFromList(util.KubeAllNamespacesFlag, ns); err != nil {
+			metrics.SendErrorLogAndMetric(util.NSID, "[UninitAllNsList] Error: failed to delete namespace set %s from list %s with err: %v", ns, util.KubeAllNamespacesFlag, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// syncAddNameSpace handles adding namespace to ipset.
+func (nsc *nameSpaceController) syncAddNameSpace(nsObj *corev1.Namespace) error {
+	var err error
+
+	nsName, nsLabel := util.GetNSNameWithPrefix(nsObj.ObjectMeta.Name), nsObj.ObjectMeta.Labels
+	log.Logf("NAMESPACE CREATING: [%s/%v]", nsName, nsLabel)
+
+	ipsMgr := nsc.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	// Create ipset for the namespace.
+	if err = ipsMgr.CreateSet(nsName, []string{util.IpsetNetHashFlag}); err != nil {
+		metrics.SendErrorLogAndMetric(util.NSID, "[AddNamespace] Error: failed to create ipset for namespace %s with err: %v", nsName, err)
```

Comment:
It may be better to consolidate this metric error to "syncNameSpace" (applied for others as well).

Reply:
this as we discussed we will be improving in the next PR which uses NpmError